modules/utxo-lib/src/bitgo/UtxoPsbt.ts

```diff
@@ -317,3 +376,145 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint>> extends Psbt {
     return { hash, sighashType };
   }
 }
+
+/**
+ * Takes a partially signed transaction and removes the scripts and signatures.
+ *
+ * Inputs must be one of:
+ *  * P2PKH
+ *  * P2SH 2-of-3
+ *  * P2WSH 2-of-3
+ *  * P2WPKH
+ *  * P2SH -> P2WSH 2-of-3
+ *  * P2SH -> P2WPKH
+ *  * P2TR script path 2-of-2
+ *
+ * @param tx the partially signed transaction
+ *
+ * @return the removed scripts and signatures, ready to be added to a PSBT
+ */
+export function unsign(tx: UtxoTransaction<bigint>, prevOuts: TxOutput<bigint>[]): PsbtInputUpdate[] {
+  const ret = tx.ins.map((input, vin) => {
+    const prevOut = prevOuts[vin];
+    const update: PsbtInputUpdate = {};
+    let redeemScript;
+    if (input.script && input.script.length) {
+      const decompiledScriptSig = bscript.decompile(input.script);
+      if (!decompiledScriptSig) {
+        throw new Error('Invalid scriptSig, failed to decompile');
+      }
+      if ((!input.witness || !input.witness.length) && decompiledScriptSig.length === 2) {
+        // P2PKH
+        const pubkey = decompiledScriptSig.pop();
+        const signature = decompiledScriptSig.pop();
+        if (!Buffer.isBuffer(pubkey) || !Buffer.isBuffer(signature)) {
+          throw new Error('Invalid pubkey or signature');
+        }
+        update.partialSig = [{ pubkey, signature }];
+        return update;
+      }
+      redeemScript = decompiledScriptSig.pop();
+      if (!Buffer.isBuffer(redeemScript)) {
+        throw new Error('Invalid redeem script');
+      }
+      update.redeemScript = redeemScript;
+      if (opcodes.OP_0 !== redeemScript[0] || ![22, 34].includes(redeemScript.length)) {
+        // P2SH -> 2-of-3
+        const hashFn = (hashType) => tx.hashForSignature(vin, redeemScript, hashType, prevOut.value);
+        update.partialSig = matchSignatures(hashFn, redeemScript, decompiledScriptSig);
+        return update;
+      }
+    }
+    if (input.witness && input.witness.length) {
+      const witness = [...input.witness];
```

Comment:
I believe this clones `input.witness`, in which case when we do `witness.pop()` it's not actually removing anything from `input.witness` if I follow correctly. But don't we want to remove the scripts and sigs from the tx within this method?

Reply:
TBH, I forget why I ended up cloning the witness here, but I know it ended up being cleaner, and I changed to clearing the witness later: https://github.com/BitGo/BitGoJS/pull/2807/files#diff-e26e7163241b590c68c95bdc501223630eb5d85391772dcfe31ccf1bc11a9589R492

I'll try to brain spelunk and add a comment.

Reply:
This turned out to be concealing a bug in `UtxoTransaction.clone`. Unbugged, and uncloned here.